File: discipline_ts/core/chronic/date.ts
import { Branded, Duration, FailureCode, TextualError, Tried } from "../x.ts";

const brand = Symbol();

export type Date = Branded<typeof brand, {
  date: globalThis.Date,
}>;

export const construct = (date: globalThis.Date): Date => {
  return Branded(brand, { date });
};

export const MINIMUM_TIMESTAMP = -8.64e15;
export const MAXIMUM_TIMESTAMP = 8.64e15;

export const now = (): Date => {
  const date = new globalThis.Date();
  date.setUTCHours(0, 0, 0, 0);
  return construct(date);
};

export const getTimestamp = (it: Date): number => {
  return it.date.getTime();
};

export const tillOrZero = (it: Date, rhs: Date): Duration => {
  const lhsTimestamp = it.date.getTime();
  const rhsTimestamp = rhs.date.getTime();

  if (lhsTimestamp < rhsTimestamp) {
    return Duration.fromMillisecondsOrThrow(rhsTimestamp - lhsTimestamp);
  } else {
    return Duration.zero();
  }
};

export const sinceOrZero = (it: Date, rhs: Date): Duration => {
  const lhsTimestamp = it.date.getTime();
  const rhsTimestamp = rhs.date.getTime();

  if (lhsTimestamp > rhsTimestamp) {
    return Duration.fromMillisecondsOrThrow(lhsTimestamp - rhsTimestamp);
  } else {
    return Duration.zero();
  }
};

export const fromTimestamp = (timestamp: number): Tried<Date, TextualError> => {
  if (!Number.isInteger(timestamp)) {
    const it = TextualError.create("Creating a Date from a millisecond timestamp since the unix epoch");
    TextualError.addMessage(it, "Argument 'timestamp' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    return Tried.Failure(it);
  }

  if (timestamp < MINIMUM_TIMESTAMP) {
    const it = TextualError.create("Creating a Date from a millisecond timestamp since the unix epoch");
    TextualError.addMessage(it, "Argument 'timestamp' is less than the minimum value");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    TextualError.addNumberAttachment(it, "Minimum value", MINIMUM_TIMESTAMP);
    return Tried.Failure(it);
  }
  
  if (timestamp > MAXIMUM_TIMESTAMP) {
    const it = TextualError.create("Creating a Date from a millisecond timestamp since the unix epoch");
    TextualError.addMessage(it, "Argument 'timestamp' is greater than the minimum value");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    TextualError.addNumberAttachment(it, "Maximum value", MAXIMUM_TIMESTAMP);
    return Tried.Failure(it);
  }

  const date = new globalThis.Date(timestamp);

  if (Number.isNaN(date.getTime())) {
    const it = TextualError.create("Creating a Date from a millisecond timestamp since the unix epoch");
    TextualError.addMessage(it, "Argument 'timestamp' is valid, but failed to produce a valid JavaScript Date");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    return Tried.Failure(it);
  }

  if (
    date.getUTCHours() !== 0
    ||
    date.getUTCMinutes() !== 0
    ||
    date.getUTCSeconds() !== 0
    ||
    date.getUTCMilliseconds() !== 0
  ) {
    const it = TextualError.create("Creating a Date from a millisecond timestamp since the unix epoch");
    TextualError.addMessage(it, "Argument 'timestamp' is valid, but produced a JavaScript Date with a non-zero time");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    TextualError.addStringAttachment(it, "JavaScript Date", date.toISOString());
    return Tried.Failure(it);
  }

  return Tried.Success(construct(date));
};

export const fromTimestampOrError = (
  timestamp: number,
  textualError: TextualError,
): Date | FailureCode => {

};

export const toString = (it: Date): string => {
  return it.date.toISOString();
};

export const getDayStart = (it: Date): Date => {
  const clone = new globalThis.Date(it.date);
  clone.setUTCHours(24, 0, 0, 0);
  return construct(clone);
};

export const getDurationTillMidnight = (it: Date): Duration => {
  const clone = new globalThis.Date(it.date);
  clone.setUTCHours(24, 0, 0, 0);

  const fromTimestamp = it.date.getTime();
  const tillTimestamp = clone.getTime();
  const difference = tillTimestamp - fromTimestamp;

  return Duration.fromMillisecondsOrThrow(difference);
};

export const isLaterThan = (it: Date, rhs: Date): boolean => {
  return getTimestamp(it) > getTimestamp(rhs);
};

export const Date = {
  MINIMUM_TIMESTAMP,
  MAXIMUM_TIMESTAMP,
  construct,
  now,
  getTimestamp,
  tillOrZero,
  sinceOrZero,
  fromTimestamp,
  fromTimestampOrError,
  toString,
  getDayStart,
  getDurationTillMidnight,
  isLaterThan,
};

File: discipline4/app/src/main/java/com/example/app/chronic/Date.kt
// package com.example.app

// import arrow.core.Either
// import arrow.core.raise.either
// import arrow.core.left
// import arrow.core.right
// import java.time.Instant
// import java.time.LocalDate
// import java.time.ZoneOffset
// import com.example.app.Duration
// import com.example.app.DateTime

// /**
//  * Represents a date (no time component) at UTC midnight
//  */
// @JvmInline
// value class Date private constructor(private val timestamp: Long) {
  
//   companion object {
//     const val MINIMUM_TIMESTAMP = -8_640_000_000_000_000L
//     const val MAXIMUM_TIMESTAMP = 8_640_000_000_000_000L
    
//     /**
//      * Creates a Date from a timestamp (must be at UTC midnight)
//      */
//     fun fromTimestamp(timestamp: Long): Either<TextualError, Date> = either {
//       if (timestamp < MINIMUM_TIMESTAMP) {
//         TextualError.create("Creating a Date from a millisecond timestamp since the Unix epoch")
//           .addMessage("Argument 'timestamp' is less than the minimum value")
//           .addNumberAttachment("Argument 'timestamp'", timestamp.toDouble())
//           .addNumberAttachment("Minimum value", MINIMUM_TIMESTAMP.toDouble())
//           .left()
//           // .bind()
//       }
      
//       if (timestamp > MAXIMUM_TIMESTAMP) {
//         TextualError.create("Creating a Date from a millisecond timestamp since the Unix epoch")
//           .addMessage("Argument 'timestamp' is greater than the maximum value")
//           .addNumberAttachment("Argument 'timestamp'", timestamp.toDouble())
//           .addNumberAttachment("Maximum value", MAXIMUM_TIMESTAMP.toDouble())
//           .left()
//           // .bind()
//       }
      
//       // Verify it's at midnight UTC
//       val instant = Instant.ofEpochMilli(timestamp)
//       val localDateTime = instant.atZone(ZoneOffset.UTC)
      
//       if (localDateTime.hour != 0 || localDateTime.minute != 0 || localDateTime.second != 0 || localDateTime.nano != 0) {
//         TextualError.create("Creating a Date from a millisecond timestamp since the Unix epoch")
//           .addMessage("Argument 'timestamp' produced a date with a non-zero time component")
//           .addNumberAttachment("Argument 'timestamp'", timestamp.toDouble())
//           .addStringAttachment("DateTime", localDateTime.toString())
//           .left()
//           // .bind()
//       }
      
//       Date(timestamp).right().bind()
//     }
    
//     fun fromLocalDate(localDate: LocalDate): Date = 
//       Date(localDate.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli())
    
//     fun now(): Date {
//       return fromLocalDate(LocalDate.now(ZoneOffset.UTC))
//     }
    
//     fun fromDateTime(dateTime: DateTime): Date = dateTime.getDate()
//   }
  
//   fun toTimestamp(): Long = timestamp
  
//   fun toLocalDate(): LocalDate = Instant.ofEpochMilli(timestamp).atZone(ZoneOffset.UTC).toLocalDate()
  
//   fun tillOrZero(other: Date): Duration {
//     return if (timestamp < other.timestamp) {
//       Duration.fromMillisecondsOrThrow(other.timestamp - timestamp)
//     } else {
//       Duration.zero()
//     }
//   }
  
//   fun sinceOrZero(other: Date): Duration {
//     return if (timestamp > other.timestamp) {
//       Duration.fromMillisecondsOrThrow(timestamp - other.timestamp)
//     } else {
//       Duration.zero()
//     }
//   }
  
//   fun isLaterThan(other: Date): Boolean {
//     return timestamp > other.timestamp
//   }

//   fun isEarlierThan(other: Date): Boolean {
//     return timestamp < other.timestamp
//   }

//   fun isAt(other: Date): Boolean {
//     return timestamp == other.timestamp
//   }
  
//   override fun toString(): String {
//     return toLocalDate().toString()
//   }
  
//   operator fun plus(days: Long): Date {
//     return Date(timestamp + days * Duration.MILLISECONDS_PER_DAY)
//   }
  
//   operator fun minus(days: Long): Date {
//     return Date(timestamp - days * Duration.MILLISECONDS_PER_DAY)
//   }

//   operator fun compareTo(other: Date): Int {
//     return timestamp.compareTo(other.timestamp)
//   }
// }